mxfusion/inference/inference_alg.py

```diff
@@ -28,28 +29,47 @@ class ObjectiveBlock(HybridBlock):
     """
     def __init__(self, infr_method, constants, data_def, var_trans, var_ties,
                  excluded, prefix='', params=None):
-        super(ObjectiveBlock, self).__init__(prefix=prefix, params=params)
+        super(ObjectiveBlock, self).__init__(prefix=prefix, params=params.param_dict)
         self._infr_method = infr_method
         self._constants = constants
         self._data_def = data_def
         self._var_trans = var_trans
         self._var_ties = var_ties
-        for name in params:
+        self._infr_params = params
+        for name in params.param_dict:
             if name not in excluded:
-                setattr(self, name, params.get(name))
+                setattr(self, name, params.param_dict.get(name))
 
     def hybrid_forward(self, F, x, *args, **kw):
-
+        """
+        This function does all the preprocesses and postprocesses for the execution of a InferenceAlgorithm.
+
+        :param F: the MXNet computation mode
+        :type F: mxnet.symbol or mxnet.ndarray
+        :param x: a dummy variable to enable the execution of this Gluon block
+        :type x: MXNet NDArray or MXNet Symbol
+        :param *arg: all the positional arguments, which correspond to the data provided to the InferenceAlgorithm.
+        :type *arg: list of MXNet NDArray or MXNet Symbol
+        :parma **kw: all the keyword arguments, which correspond to the parameters that may require gradients.
+        :type kw: {str(UUID): MXNet NDArray or MXNet Symbol}
+        :returns: the outcome of the InferenceAlgorithm that are determined by the inference algorithm.
+        :rtypes: {str: MXNet NDArray or MXNet Symbol}
+        """
         for to_uuid, from_uuid in self._var_ties.items():
             kw[to_uuid] = kw[from_uuid]
         data = {k: v for k, v in zip(self._data_def, args)}
-        data = add_sample_dimension_arrays(F, data)
+        variables = add_sample_dimension_to_arrays(F, data)
         for k, v in self._var_trans.items():
             kw[k] = v.transform(kw[k], F=F)
-        kw = add_sample_dimension_arrays(F, kw)
-        constants = add_sample_dimension_arrays(F, self._constants)
-        obj = self._infr_method.compute(F=F, data=data, parameters=kw,
-                                        constants=constants)
+        add_sample_dimension_to_arrays(F, kw, out=variables)
+        add_sample_dimension_to_arrays(F, self._constants, out=variables)
+        obj = self._infr_method.compute(F=F, variables=variables)
+        with autograd.pause():
+            # An inference algorithm may directly set the value of a parameter instead of computing its gradient.
+            # This part handles the setting of parameters.
+            for k, v in variables.items():
+                if k.startswith('SET_'):
```

Comment:
It is difficult because *variables* is the only argument that has been passed around during execution. This set_parameter function may be called by an internal inference algorithm in the case module, which does not have access to the InferenceParameter.